### DISCUSSION TITLE
tower_http::ServeDir no longer compatible with axum 0.7

### DISCUSSION THREAD

**Discussion by slatian** [tokio-rs/axum#2418] [Q&A] [answered]
- [x] I have looked for existing issues (including closed) about this

## Bug Report

In previous versions (0.6 series) of axum one could use the very useful [tower_http::ServeDir](https://docs.rs/tower-http/0.5.0/tower_http/services/struct.ServeDir.html) with an axum Router (other types are affected as well), simply by adding it as a service.

```rust
let app = Router::new()
    .route("/", get(handle_default_route))
    .fallback_service(
        ServeDir::new(static_file_directory)
            .fallback(not_found_handler.with_state(shared_state.clone()))
    )
```

This no longer works because (I aassume) the responses are now `Response<Body>` instead of being able to be a `Response<UnsyncBoxBody<Bytes, BoxError>` which is still used by the `tower_http`.

> breaking: The following types/traits are no longer generic over the request body
(i.e. the B type param has been removed) (https://github.com/tokio-rs/axum/pull/1751 and https://github.com/tokio-rs/axum/pull/1789):
> […]

I have not seen anyone solve this for axum 0.7 and was unable to solve this myself.

Maybe there is a simple solution in which case official documentation on how to make this work again would be nice.

**Comment by jplatte** [comment, marked as answer]
Please make sure to also upgrade tower-http to 0.5.0. The old versions of axum and tower-http used the request and response types from v0.2.x of the http create, while the new ones have upgraded to v1.x.

**Reply by slatian** [reply]
Little details, always … thank you!
